### PR TITLE
Try to introduce TSystem::GetCryptoRandom

### PR DESCRIPTION
**Pull request by linev**
# This Pull request:

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

File: core/unix/src/TUnixSystem.cxx [hunk 1]
```diff
@@ -724,6 +725,16 @@ const char *TUnixSystem::GetError()
...
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Return cryptographic random number
+/// Fill provided buffer with random values
+/// Returns number of bytes written to buffer or -1 in case of error
+
+Int_t TUnixSystem::GetCryptoRandom(void *buf, Int_t len)
+{
+   return getrandom(buf, len, GRND_RANDOM | GRND_NONBLOCK);
```
[the remark this answers is not in the document]
Comment (linev) [reply core/unix/src/TUnixSystem.cxx:735]: Ok, I will rebase PR  - after windows build will be done

**Review by bellenot — APPROVED** [review]
The Windows part looks good to me, if the builds are successful 😉

**Comment by linev** [comment]
@jblomer Can you check now?

**Review by jblomer — COMMENTED** [review]
Looks good to me except for macOS, which apparently should use `arc4random_buf()`. Or perhaps just omit the `GRND_NONBLOCK` flag on macOS.

**Comment by linev** [comment]
> Looks good to me except for macOS, which apparently should use arc4random_buf().

I prefer to use same method on Mac - just excluding flag which is not supported.

**Comment by linev** [comment]
@jblomer @bellenot 

`arc4random_buf()` looks like preferable method to get cryptographic random numbers and on newer Linux it provided with `<stdlib.h>`.
But on Mac or Ubuntu or alma it is not a case. 
On some platforms (including Mac) one can use `libbsd` - but again not always.
And finally there is `read_random` function. 

Seems to be, one need to test all them one after another with `CHECK_CXX_SOURCE_COMPILES` to be able use it.

**Comment by bellenot** [comment]
@linev I could check how to do that, but isn't there any simpler solution?

**Comment by linev** [comment]
@bellenot 

> isn't there any simpler solution

I have no idea. While it is not real standard, each linux/mac flawor has own way to provide function to read `/dev/random` or 
`/dev/urandom`

**Comment by linev** [comment]
And I already preparing new commit for this.

**Comment by linev** [comment]
@jblomer 

Can you check again?
I using now `arc4random_buf` by default, but trying some other options.

File: core/unix/CMakeLists.txt [hunk 2]
```diff
@@ -20,4 +20,29 @@ if (CMAKE_SYSTEM_NAME MATCHES FreeBSD)
...
+     message(STATUS "Found arc4random_buf in bsd/stdlib.h")
+     target_compile_definitions(Core PRIVATE R__ARC4_BSDLIB)
+  else()
+    CHECK_CXX_SOURCE_COMPILES("#include <sys/random.h>
+    int main() { char buf[32]; int res = getrandom(buf, 32,  GRND_RANDOM | GRND_NONBLOCK); return 0;}" found_getrandom)
+    if(found_getrandom)
+      message(STATUS "Found getrandom in sys/random.h")
+      target_compile_definitions(Core PRIVATE R__GETRANDOM_CLIB)
+    else()
+      message(FATAL_ERROR "Fail to detect reliable random generator")
```
Comment (jblomer) [inline core/unix/CMakeLists.txt:42]:
```suggestion
      message(FATAL_ERROR "Fail to detect cryptographic random generator")
```

File: core/unix/src/TUnixSystem.cxx [hunk 3]
```diff
@@ -724,6 +732,24 @@ const char *TUnixSystem::GetError()
...
+/// Fill provided buffer with random values
+/// Returns number of bytes written to buffer or -1 in case of error
+
+Int_t TUnixSystem::GetCryptoRandom(void *buf, Int_t len)
+{
+   #if defined(R__ARC4_STDLIB) || defined(R__ARC4_BSDLIB)
+   arc4random_buf(buf, len);
+   return len;
+   #elif defined(R__GETRANDOM_CLIB)
+   return getrandom(buf, len,  GRND_RANDOM | GRND_NONBLOCK);
```
Comment (jblomer) [inline core/unix/src/TUnixSystem.cxx:746]:
```suggestion
   return getrandom(buf, len, GRND_NONBLOCK);
```

**Review by jblomer — COMMENTED** [review]
The approach looks good to me.

There are still test errors and clang-format errors to handle.

File: core/unix/src/TUnixSystem.cxx [hunk 4]
```diff
@@ -724,6 +732,24 @@ const char *TUnixSystem::GetError()
...
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Return cryptographic random number
+/// Fill provided buffer with random values
+/// Returns number of bytes written to buffer or -1 in case of error
+
+Int_t TUnixSystem::GetCryptoRandom(void *buf, Int_t len)
+{
+   #if defined(R__ARC4_STDLIB) || defined(R__ARC4_BSDLIB)
```
Comment (bellenot) [inline core/unix/src/TUnixSystem.cxx:742]:
```suggestion
#if defined(R__ARC4_STDLIB) || defined(R__ARC4_BSDLIB)
```

File: core/unix/src/TUnixSystem.cxx [hunk 5]
```diff
@@ -724,6 +732,24 @@ const char *TUnixSystem::GetError()
...
+   arc4random_buf(buf, len);
+   return len;
+   #elif defined(R__GETRANDOM_CLIB)
```
Comment (bellenot) [inline core/unix/src/TUnixSystem.cxx:745]:
```suggestion
#elif defined(R__GETRANDOM_CLIB)
```

File: core/unix/src/TUnixSystem.cxx [hunk 6]
```diff
@@ -724,6 +732,24 @@ const char *TUnixSystem::GetError()
...
+   return getrandom(buf, len,  GRND_RANDOM | GRND_NONBLOCK);
+   #else
+   #error "Reliable cryptographic random function not defined"
+   return -1;
+   #endif
```
Comment (bellenot) [inline core/unix/src/TUnixSystem.cxx:750]:
```suggestion
#endif
```

**Review by bellenot — APPROVED** [review]
LGTM, but see my suggestions

**Review by jblomer — APPROVED** [review]
LGTM (but see remaining clang-format issues)

**Comment by linev** [comment]
> but see remaining clang-format issues

Most of clang format warnings about header files.
We need to reformat all ROOT headers to avoid them.

**Comment by jblomer** [comment]
Right, my understanding is that we want to fix the formatting as we go, incrementally. So that over time the situation improves.

**Comment by linev** [comment]
> So that over time the situation improves.

It doing this on PR basis we will break formatting of header files - different lines will have different format.
One should do it once for all headers and probably for all source files.

**Pull request merged**